### DISCUSSION TITLE
Can't Split edges into pos and neg edge_index

### DISCUSSION THREAD

**Discussion by atale480** [pyg-team/pytorch_geometric#6344] [Q&A] [answered]
I'm trying to build Adversarially Regularised Graph Auto-Encoder. For this I'm using [Princeton Segmentation Benchmark](https://segeval.cs.princeton.edu) as dataset. When I look at the documentation of PyTorch_Grometric for Adversarially Regularised Graph Auto-Encoder I came to know I need pos_edge_index and neg_edge_index. For this I thought of using `torch_geometric.utils.train_test_splitedges`method. But As this method is going to be deprecated soon and as mentioned in that documentation I'm using `torch_geometric.transforms.RandomLinkSplit` method. But when I tried to used this method I got following error. 

Here's code I wrote
```
dataset = MyOwnDataset(root = 'data/MeshData/')
dataset = dataset.shuffle()
transform = RandomLinkSplit(is_undirected=True)
train_data, val_data, test_data = transform(dataset)
```
Error I got
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
Input In [11], in <cell line: 2>()
      1 transform = RandomLinkSplit(is_undirected=True)
----> 2 train_data, val_data, test_data = transform(dataset)

File ~/miniforge3/envs/pytorch/lib/python3.8/site-packages/torch_geometric/transforms/random_link_split.py:140, in RandomLinkSplit.__call__(self, data)
    138 else:
    139     rev_edge_types = [None]
--> 140     stores = [data._store]
    141     train_stores = [train_data._store]
    142     val_stores = [val_data._store]

AttributeError: 'MyOwnDataset' object has no attribute '_store'
```

I've used basic method while creating own dataset like raw_paths, raw_dir, processed_dir, raw_file_names, processed_file_names. 
I don't know what is _store attribute. Can anyone let me know what is that? What is that I missed?
Also I want to know if it is possible with decoder that I pass encoder output to decoder and decoder will output 1 dimensional vector? if yes then how to do that?

> Environment 
>- PyG version: 2.1.0.post1 
>- PyTorch Version: 1.13.0.dev20220707
>- PyTorch Scatter Version: 2.0.9
>- PyTorch Sparse Version: 0.6.14
>- OS: MacOS Ventura 13.1
>- Python Version: 3.8.13
>- Installation Source( `conda, pip`): pip

**Comment by wsad1** [comment, marked as answer]
`RandomLinkSplit` needs a `Data` , which is a single graph, object as input. 
```
dataset = MyOwnDataset(root = 'data/MeshData/')
dataset = dataset.shuffle()
from torch_geometric.data import Batch
# Convert from n graphs to one large graph with n disconnected components
data = Batch.from_data_list(dataset[:])
transform = RandomLinkSplit(is_undirected=True)
train_data, val_data, test_data = transform(dataset)
```

**Reply by atale480** [reply]
`# Convert from n graphs to one large graph with n disconnected components`
How to do this?

**Reply by atale480** [reply]
Is is possible to create Adversarially Regularised Graph Auto-Encoder without splitting data in pos and neg edge index? If yes how to do that?

**Reply by rusty1s** [reply]
The `Batch.from_data_list` converts the graphs to one large graph with disconnected components. Not sure what you mean with your second point: you will always need positive and negative links/examples in this case.

**Reply by atale480** [reply]
@rusty1s as per your suggestion I used `Batch.from_data_list` I wrote following code and got some output.
**Code I Wrote**
```
dataset = MyOwnDataset(root = 'data/MeshData/')
dataset = dataset.shuffle()
batch = Batch.from_data_list(dataset)
transform = RandomLinkSplit(num_val=0.1, num_test=0.1 , is_undirected=True)
train_data, val_data, test_data = transform(batch)
print(batch)
print(train_data)
print(val_data)
print(test_data)
```

**Output I got**
```
DataBatch(x=[3885011, 3], edge_index=[2, 23306574], y=[380], batch=[3885011], ptr=[381])
DataBatch(x=[3885011, 3], edge_index=[2, 18645262], y=[380], batch=[3885011], ptr=[381], edge_label=[18645262], edge_label_index=[2, 18645262])
DataBatch(x=[3885011, 3], edge_index=[2, 18645262], y=[380], batch=[3885011], ptr=[381], edge_label=[2330656], edge_label_index=[2, 2330656])
DataBatch(x=[3885011, 3], edge_index=[2, 20975918], y=[380], batch=[3885011], ptr=[381], edge_label=[2330656], edge_label_index=[2, 2330656])
```

I can't see pos and neg edge_index. Can you guid me for creating ARGA.

Also I tried to use data loader then I was getting error when I tried to print it
**Code for data loader**
```
train_loader = DataLoader(train_data, batch_size=64)
val_loader   = DataLoader(val_data, batch_size=64)
test_loader  = DataLoader(test_data, batch_size=64)

for i in train_loader:
    print(i)
```

**Output I got**
<img width="1324" alt="Screenshot 2023-01-06 at 12 48 05" src="https://user-images.githubusercontent.com/40666896/210950494-01bfc7e2-3d92-4b56-a55f-6fb83a6da584.png">
<img width="1324" alt="Screenshot 2023-01-06 at 12 48 30" src="https://user-images.githubusercontent.com/40666896/210950521-1a07e5d1-151c-42d0-b2db-dc2905f98702.png">

**Reply by rusty1s** [reply]
Sorry for the confusion. It looks like I was wrong: ÀRGVA` doesn't need any negative edges. [Here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/argva_node_clustering.py) is a full example on how to use. All you need to change is iterate over your data in mini-batches.

**Reply by atale480** [reply]
I followed that example but got stuck
Code
```
import torch_geometric.transforms as T
transform = T.Compose([T.RandomLinkSplit(num_val=0.1, num_test=0.1, is_undirected=True,
                                         split_labels=True, add_negative_train_samples=False),])
dataset = MyOwnDataset(root = 'data/MeshData/', transform=transform)
dataset = dataset.shuffle()
print(dataset[2])
train_data, val_data, test_data = dataset
```

Output
```
(Data(x=[5614, 3], edge_index=[2, 26940], y=[1], pos_edge_label=[13470], pos_edge_label_index=[2, 13470]),
 Data(x=[5614, 3], edge_index=[2, 26940], y=[1], pos_edge_label=[1683], pos_edge_label_index=[2, 1683], neg_edge_label=[1683], neg_edge_label_index=[2, 1683]),
 Data(x=[5614, 3], edge_index=[2, 30306], y=[1], pos_edge_label=[1683], pos_edge_label_index=[2, 1683], neg_edge_label=[1683], neg_edge_label_index=[2, 1683]))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Input In [15], in <cell line: 5>()
      3 dataset = MyOwnDataset(root = 'data/MeshData/', transform=transform)
      4 dataset = dataset.shuffle()
----> 5 train_data, val_data, test_data = dataset

ValueError: too many values to unpack (expected 3)
```
As you see this is creating 3 data object for each graph. If I try to use `train_data, val_data, test_data = transform(batch)` then again I get error that dataset has no attribute _store

When I get data in train, val and test with pos and neg edge_label_index then only I can use data loader and then iterate model in mini-batches. Also I tried following code for data loader after getting above error then I did not get pos and neg edge_label_index in that when I tried to print DataLoader.
```
train_dataset = dataset[:int(len(dataset)*0.8)]
test_dataset = dataset[int(len(dataset)*0.1):]

train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False)
```

Maybe I'm not able to understand what is written GitHub example.  **But Question I'm getting is that how to split data in train test test and val for given batch size along with pos and neg edge_label_index of each (i.e. pos and neg edge_label_index for train split, val split and test split)**

**Reply by atale480** [reply]
> I followed that example but got stuck Code
> 
> ```
> import torch_geometric.transforms as T
> transform = T.Compose([T.RandomLinkSplit(num_val=0.1, num_test=0.1, is_undirected=True,
>                                          split_labels=True, add_negative_train_samples=False),])
> dataset = MyOwnDataset(root = 'data/MeshData/', transform=transform)
> dataset = dataset.shuffle()
> print(dataset[2])
> train_data, val_data, test_data = dataset
> ```
> 
> Output
> 
> ```
> (Data(x=[5614, 3], edge_index=[2, 26940], y=[1], pos_edge_label=[13470], pos_edge_label_index=[2, 13470]),
>  Data(x=[5614, 3], edge_index=[2, 26940], y=[1], pos_edge_label=[1683], pos_edge_label_index=[2, 1683], neg_edge_label=[1683], neg_edge_label_index=[2, 1683]),
>  Data(x=[5614, 3], edge_index=[2, 30306], y=[1], pos_edge_label=[1683], pos_edge_label_index=[2, 1683], neg_edge_label=[1683], neg_edge_label_index=[2, 1683]))
> ---------------------------------------------------------------------------
> ValueError                                Traceback (most recent call last)
> Input In [15], in <cell line: 5>()
>       3 dataset = MyOwnDataset(root = 'data/MeshData/', transform=transform)
>       4 dataset = dataset.shuffle()
> ----> 5 train_data, val_data, test_data = dataset
> 
> ValueError: too many values to unpack (expected 3)
> ```
> 
> As you see this is creating 3 data object for each graph. If I try to use `train_data, val_data, test_data = transform(batch)` then again I get error that dataset has no attribute _store
> 
> When I get data in train, val and test with pos and neg edge_label_index then only I can use data loader and then iterate model in mini-batches. Also I tried following code for data loader after getting above error then I did not get pos and neg edge_label_index in that when I tried to print DataLoader.
> 
> ```
> train_dataset = dataset[:int(len(dataset)*0.8)]
> test_dataset = dataset[int(len(dataset)*0.1):]
> 
> train_loader = DataLoader(train_dataset, batch_size=64, shuffle=True)
> test_loader = DataLoader(test_dataset, batch_size=64, shuffle=False)
> ```
> 
> Maybe I'm not able to understand what is written GitHub example. **But Question I'm getting is that how to split data in train test test and val for given batch size along with pos and neg edge_label_index of each (i.e. pos and neg edge_label_index for train split, val split and test split)**

My dataset contain 380 graphs and I want to divide 380 graph in train test and val with some batch size and along with this I need edge_label_index for each.

**Reply by rusty1s** [reply]
I think you would need to call this via `train_data, val_data, test_data = zip(dataset)`.

**Reply by atale480** [reply]
Thanks for support. 
I wrote simple code to get train, val and test data as follows
```
train_data = []
val_data = []
test_data = []
for i in dataset:
    a,b,c = i
    train_data.append(a)
    val_data.append(b)
    test_data.append(c)
```
